remote-drivers/lwip-ethernet-async/client.c

```diff
@@ -0,0 +1,446 @@
+ * @param buffer ethernet buffer to return. 
+ *
+ */
+static inline void return_buffer(state_t *state, ethernet_buffer_t *buffer)
+{
+    assert(buffer->origin == ORIGIN_RX_QUEUE);
+    ZF_LOGW("Returning RX buffer");
+    /* As the rx_available ring is the size of the number of buffers we have, 
+    the ring should never be full. */
+    enqueue_avail(state->rx_ring, buffer->dma_addr, BUF_SIZE, buffer);
+}
+
+/**
+ * Allocate an empty TX buffer from the empty pool
+ *
+ * @param state client state data. 
+ * @param length length of buffer required 
+ *
+ */
+static inline ethernet_buffer_t *alloc_tx_buffer(state_t *state, size_t length)
+{   
+    if (BUF_SIZE < length) {
+        ZF_LOGE("Requested buffer size too large.");
+        return NULL;
+    }
+
+    uintptr_t encoded_addr;
+    size_t len;
+    ethernet_buffer_t *buffer;
+
+    int err = dequeue_avail(state->tx_ring, &encoded_addr, &len, (void **)&buffer);
+    ZF_LOGF_IF(err, "tx available ring is empty");
+
+    /* Little sanity check */
+    if (buffer->dma_addr != encoded_addr) {
+        ZF_LOGE("buffer->dma_addr != encoded_addr, dma_addr = %p, encoded_addr = %p", buffer->dma_addr, encoded_addr);
+    }
+
+    return buffer;
+}
+
+typedef struct lwip_custom_pbuf {
+    struct pbuf_custom custom;
+    ethernet_buffer_t *buffer;
+    state_t *state;
+} lwip_custom_pbuf_t;
+LWIP_MEMPOOL_DECLARE(
+    RX_POOL,
+    NUM_BUFFERS * 2,
+    sizeof(lwip_custom_pbuf_t),
+    "Zero-copy RX pool"
+);
+
+/**
+ * Free a pbuf. This also returns the underlying buffer to
+ * the appropriate place. 
+ *
+ * @param buf pbuf to free. 
+ *
+ */
+static void interface_free_buffer(struct pbuf *buf)
+{
+    SYS_ARCH_DECL_PROTECT(old_level);
+
+    lwip_custom_pbuf_t *custom_pbuf = (lwip_custom_pbuf_t *) buf;
+
+    SYS_ARCH_PROTECT(old_level);
+    return_buffer(custom_pbuf->state, custom_pbuf->buffer);
+    LWIP_MEMPOOL_FREE(RX_POOL, custom_pbuf);
+    SYS_ARCH_UNPROTECT(old_level);
+}
+
+/**
+ * Create a pbuf structure to pass to the network interface. 
+ *
+ * @param state client state data. 
+ * @param buffer ethernet buffer containing metadata for the actual buffer
+ * @param length length of data
+ * 
+ * @return the newly created pbuf. 
+ */
+static struct pbuf *create_interface_buffer(state_t *state, ethernet_buffer_t *buffer, size_t length) 
+{
+    lwip_custom_pbuf_t *custom_pbuf = (lwip_custom_pbuf_t *) LWIP_MEMPOOL_ALLOC(RX_POOL);
+
+    custom_pbuf->state = state;
+    custom_pbuf->buffer = buffer;
+    custom_pbuf->custom.custom_free_function = interface_free_buffer;
+
+    return pbuf_alloced_custom(
+        PBUF_RAW,
+        length,
+        PBUF_REF,
+        &custom_pbuf->custom,
+        buffer->buffer,
+        buffer->size
+    );
+}
+
+/**
+ * Packets have been received and are ready for processing. 
+ * Dequeue the packets from shared memory and pass them to lwip to proess. 
+ *
+ * @param cookie client state data. 
+ *
+ */
+static void rx_queue(void *cookie)
```

Comment:
Could make the name a bit more self-explaining:
```suggestion
static void process_rx_queue(void *cookie)
```